### DISCUSSION TITLE
Help with solving conflicts caused by analogous syntax

### DISCUSSION THREAD

**Discussion by P233** [tree-sitter/tree-sitter#2199] [Q&A] [answered]
Hi. I am new to tree-sitter and am currently trying to create a new parser for SCSS syntax. However, I have encountered a problem and am unsure how to fix it. 

The SCSS `list` and `map` syntax are very similar, for example:

```scss
$map: (
  width: 20px,
  height: $var
);

$list: (1 2 3);
```

Both of them are wrapped in `()` and contain multiple entries. The `map` entry is a key-value pair, and the `list` entry is a value.  Here is a part of my parser code:

```js
list: $ => {
  return choice(
    seq("(", repeat(seq($._list_value, optional(","))), ")"),
    seq("[", repeat(seq($._list_value, optional(","))), "]")
  );
},

map: $ => {
  return seq("(", repeat($.map_entry), ")");
},
```

The list and map are definitely conflicting. I am unsure how to make tree-sitter use the correct rule for each syntax, because it either use `list` or `map` to match both of the two code snippets. Not sure if it has anything to do with precedences.

**Comment by ahelwer** [comment]
Given an example input, what is your desired parse tree and what is the actual parse tree?

**Comment by P233** [comment]
Hi @ahelwer thanks for the quick reply.

Here are more of my parser code:

```js
stylesheet: $ => repeat(choice($.variable_declaration)),

variable_declaration: $ => {
  return seq($.variable_name, ":", $._variable_value, optional($.flag), ";");
},

_variable_value: $ => {
  return repeat1(
    choice(
      $.number,
      $.string,
      $.operator,
      $.hex_color,
      $.map,  // conflict with $.list
      $.list, // conflict with $.map
      $.variable_name,
      // $.interpolation,
      $.call_expression,
      $.plain_value,
      ","
    )
  );
},

list: $ => {
  return choice(
    seq("(", repeat(seq($._list_value, optional(","))), ")"),
    seq("[", repeat(seq($._list_value, optional(","))), "]")
  );
},

map: $ => {
  return seq("(", repeat($.map_entry), ")");
},
```

You could see the full code from https://github.com/P233/tree-sitter-scss/blob/main/grammar.js, this repo is still working in progress.

So, when I attempted to parse 

```scss
$map: (
  width: 20px,
  height: $var
);

$list: (1 2 3);
```

I would have:

```
(stylesheet [0, 0] - [6, 0]
  (variable_declaration [0, 0] - [3, 2]
    (variable_name [0, 0] - [0, 4])
    (list [0, 6] - [3, 1]
      (ERROR [1, 2] - [1, 8]
        (function_name [1, 2] - [1, 7]))
      (number [1, 9] - [1, 13]
        (unit [1, 11] - [1, 13]))
      (ERROR [2, 2] - [2, 9]
        (function_name [2, 2] - [2, 8]))
      (variable_name [2, 10] - [2, 14])))
  (variable_declaration [5, 0] - [5, 15]
    (variable_name [5, 0] - [5, 5])
    (list [5, 7] - [5, 14]
      (number [5, 8] - [5, 9])
      (number [5, 10] - [5, 11])
      (number [5, 12] - [5, 13]))))
./test/test.scss	0 ms	(ERROR [1, 2] - [1, 8])
```

The `$list` is parsed correctly, and the `$map` is parsed as a Sass list. But, if I comment out the `$.list` in `_variable_value`, I would have:

```
(stylesheet [0, 0] - [6, 0]
  (variable_declaration [0, 0] - [3, 2]
    (variable_name [0, 0] - [0, 4])
    (map [0, 6] - [3, 1]
      (map_entry [1, 2] - [1, 14]
        (map_key [1, 2] - [1, 7])
        (number [1, 9] - [1, 13]
          (unit [1, 11] - [1, 13])))
      (map_entry [2, 2] - [2, 14]
        (map_key [2, 2] - [2, 8])
        (variable_name [2, 10] - [2, 14]))))
  (variable_declaration [5, 0] - [5, 15]
    (variable_name [5, 0] - [5, 5])
    (map [5, 7] - [5, 14]
      (ERROR [5, 8] - [5, 13]))))
./test/test.scss	0 ms	(ERROR [5, 8] - [5, 13])
```

Then, the `$map` is parsed correctly, but the `$list` is parsed as a Sass map.

I also tried something like `choice($.map, $.list)`, it doesn't work.

The expected result is to parse the `$map` as a Sass map and the `$list` as a Sass list. However, both of them begin with `(` and end with `)`, and they can also be nested within each other. I am not quite get a handle on tree-sitter yet, so confused and frustrated.

**Reply by ahelwer** [reply, marked as answer]
It fundamentally comes down to a lexical precedence issue. Basically when parsing `width`, on the `list` side you have the `call_expression` rule which has the `function_name` regex `/[a-zA-Z][\w-]+/` while on the `map` side you have `map_key` defined as `alias(repeat1(choice(/[\w-]/, $.interpolation)), $.map_key)`. So you're pitting the repeated `/[\w-]/` regex against a single `/[a-zA-Z][\w-]+/` regex; by the longest match rule, the latter will win and so the lexer commits `width` to being a `function_name`, thus commiting the parse to be a `call_expression` within a `list`.

I suspect your definition of `map_entry` is incorrect with regard to SCSS syntax although I don't actually know the language myself. Anyway I was able to fix the issue by redefining `map_entry` as follows:

```js
    map_entry: $ => {
      return seq(
        alias(choice(/[a-zA-Z][\w-]+/, $.interpolation), $.map_key),
        ":",
        choice($.map, $._map_value),
        optional(",")
      );
    },
```
So there are no repeats for `$._map_value` nor `$.map_key`.

There are a few other practices you could adopt:
 1. Instead of having hardcoded regexes scattered throughout the grammar, centralize them in single rules for each class of identifier. This will make it easier to debug issues of this type in the future.
 2. Be wary of mixing `repeat` with regexes. This mixture of parse/lex precedence can be difficult to reason about.
 3. This is purely stylistic but rather than writing `$ => { return expr; }` you can just write `$ => expr`

This wasn't a trivial issue, it probably took me 45 minutes to debug! The intersection of lexical & parse precedence can be somewhat strange until you get a feel for how the parser works.

**Comment by P233** [comment]
Hi @ahelwer , I am truly sorry for having you spend 45 minutes debugging my issue. I really appreciate it! And thank you for the valuable suggestions, they have definitely helped me understand more about Tree-sitter.

I have had a lot of experience creating syntax highlighting with regex in the past. However, I am finding it a bit challenging to start using Tree-sitter. I'll continue learning about it and complete the project. Thank you very much for your help!

**Reply by ahelwer** [reply]
No worries, happy to provide help to those who need it, as they helped me. That's what open source is all about.
